link/periodic.go

```diff
@@ -0,0 +1,93 @@
+package link
+
+import (
+	"fmt"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/internal"
+)
+
+// PeriodicOptions defines additional parameters that will be used
+// when loading periodic samplers.
+type PeriodicOptions struct {
+	// Arbitrary value that can be fetched from an eBPF program
+	// via `bpf_get_attach_cookie()`.
+	//
+	// Needs kernel 5.15+.
+	Cookie uint64
+}
+
+// Periodic attaches the given eBPF program to a periodic perf event on given
+// cpu and given frequency.
+//
+// This is mainly useful when doing sample-based tracing or monitoring.
+//
+// Losing the reference to the resulting Link (tp) will close the Periodic event
+// and prevent further execution of prog. The Link must be Closed during
+// program shutdown to avoid leaking system resources.
+func Periodic(frequency uint64, cpu int, prog *ebpf.Program, opts *PeriodicOptions) (Link, error) {
```

Comment:
There is a number of combinations of `pid` and `cpu` that the perf subsystem accepts for periodic events:

- pid == 0 and cpu == -1
- pid == 0 and cpu >= 0
- pid > 0 and cpu == -1
- pid > 0 and cpu >= 0
- pid == -1 and cpu >= 0

This function signature only allows to set the CPU . But I do think there is a valid use case where people want to attach an eBPF program via perf only to a single PID - e.g. for debugging purposes.